### PR TITLE
[master] Build Golden Docker Containers

### PR DESCRIPTION
**Pull request by s0undt3ch**

### CODE REVIEW

**Review by bryceml — COMMENTED** [review]
The names seem a bit different from the AWS ami's.  For instance, they are capitalized and arch is called latest instead of lts.  I feel like it will be easier to automate if they have more similar names.

**Comment by s0undt3ch** [comment]
```
invoke build-docker --distro=debian --distro-version=9
```
VS
```
invoke build-aws --distro=debian --distro-version=9
```
?
So I don't see the name being a problem here.

As for Arch, it's actually not relying on any version, though we do define the version as `latest` on the JSON vars template.
Would you prefer the vars template to be renamed from `arch-vars.json` to `arch-lts-vars.json`? or `arch-latest-vars.json`?

**Comment by bryceml** [comment]
I didn't see you do `| lower` when you actually save the image.  Yeah, we will either need to change it to arch-lts-vars.json here, or we would probably end up having to change to latest all over jenkins and the AMI images.  I would probably put a version on the saved docker image too.  Still looking through it, but arch is probably the only one that needs adjusted.

**Comment by s0undt3ch** [comment]
OK. I'll change arch to `lts`

**Comment by bryceml** [comment]
Read through the rest, looks good.

**Comment by s0undt3ch** [comment]
Updated

File: Docker/amazon/amazon-1-vars.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "_comment": "Build with `packer build -var-file=amazon-1-vars.json amazon.json`",
+  "docker_image": "amazonlinux:2018.03",
```
Comment (bryceml) [inline Docker/amazon/amazon-1-vars.json]: why not amazonlinux:1 ?

File: Docker/arch/arch-lts-vars.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "_comment": "Build with `packer build -var-file=arch-vars.json arch.json`",
```
Comment (bryceml) [inline Docker/arch/arch-lts-vars.json]: this comment needs updated.

File: Docker/arch/arch-lts-vars.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+  "docker_image": "archlinux/base",
+  "docker_repository": "saltstack",
+  "docker_image_prefix": "ci",
+  "build_type": "base",
+  "os_name": "Arch",
+  "os_version": "latest",
```
Comment (bryceml) [inline Docker/arch/arch-lts-vars.json]: What is going to consume the os_version ?  do we want it to match everything else and be `lts` ?

File: Docker/arch/arch.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        "pacman -Sy --needed --noconfirm git vim",
+        "rm -rf /tmp/* /var/cache/pacman/pkg/* && pacman -Scc --noconfirm"
+      ]
+    }
+  ],
+  "post-processors": [
+    [
+      {
+        "type": "docker-tag",
+        "repository": "{{ user `docker_repository` }}/{{ user `docker_image_prefix` }}-{{ user `os_name` | lower }}",
```
Comment (bryceml) [inline Docker/arch/arch.json]: should we push lts as the version here so we match other places?

File: Docker/ubuntu/ubuntu.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        "LABEL Description=\"{{ user `build_type` | upper }} Image of {{ user `os_name` }} {{ user `os_version` }}\""
+      ],
+      "message": "{{ user `build_type` | upper }} Image of {{ user `os_name` }} {{ user `os_version` }} on {{ isotime }}"
+    }
+  ],
+  "provisioners": [
+    {
+      "type": "shell",
+      "inline": [
+        "rm -rf /etc/apt/apt.conf.d/50unattended-upgrades",
```
Comment (bryceml) [inline Docker/ubuntu/ubuntu.json]: should this be `20auto-upgrades` instead?

File: Docker/debian/debian.json [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        "LABEL Description=\"{{ user `build_type` | upper }} Image of {{ user `os_name` }} {{ user `os_version` }}\""
+      ],
+      "message": "{{ user `build_type` | upper }} Image of {{ user `os_name` }} {{ user `os_version` }} on {{ isotime }}"
+    }
+  ],
+  "provisioners": [
+    {
+      "type": "shell",
+      "inline": [
+        "rm -rf /etc/apt/apt.conf.d/50unattended-upgrades",
```
Comment (bryceml) [inline Docker/debian/debian.json]: should this be `20auto-upgrades` instead?

Comment (bryceml) [reply Docker/debian/debian.json]: https://github.com/saltstack/salt-jenkins/blob/develop/apt.sls

File: Docker/debian/files/sources.list [hunk 7]
```diff
@@ -0,0 +1,8 @@
+deb http://security.debian.org/ jessie/updates main
```
Comment (bryceml) [inline Docker/debian/files/sources.list]: this file doesn't seem to be used anywhere.

File: Docker/amazon/amazon-1-vars.json (hunk 1 above)
Comment (s0undt3ch) [reply Docker/amazon/amazon-1-vars.json]: This is the image we pull in.

Comment (bryceml) [reply Docker/amazon/amazon-1-vars.json]: yeah, https://hub.docker.com/_/amazonlinux . 1 is easier to understand what is happening.

File: Docker/ubuntu/ubuntu.json (hunk 5 above)
Comment (s0undt3ch) [reply Docker/ubuntu/ubuntu.json]: You know more about this than I do, shall I switch to what you're saying?

Comment (bryceml) [reply Docker/ubuntu/ubuntu.json]: Yeah.  We should change it for the AMI builds as well.

Comment (bryceml) [reply Docker/ubuntu/ubuntu.json]: With the AMI builds, we could either do it now or in a different pr.

File: Docker/debian/debian.json (hunk 6 above)
Comment (s0undt3ch) [reply Docker/debian/debian.json]: I fixed it.

File: Docker/ubuntu/ubuntu.json (hunk 5 above)
Comment (s0undt3ch) [reply Docker/ubuntu/ubuntu.json]: I added that change too

File: AWS/debian/debian-8.json [hunk 8]
```diff
@@ -60,7 +60,7 @@
     {
       "type": "shell",
       "inline": [
-        "sudo  rm -rf /etc/apt/apt.conf.d/50unattended-upgrades",
+        "sudo  rm -rf /etc/apt/apt.conf.d/20unattended-upgrades",
```
Comment (bryceml) [inline AWS/debian/debian-8.json]: should be 20auto-upgrades, not 20 unattended-upgrades

Comment (bryceml) [reply AWS/debian/debian-8.json]: even safer would be to 
```bash
echo "" > /etc/apt/apt.conf.d/20auto-upgrades
```
so any package updates don't re-create the file.

Comment (bryceml) [reply AWS/debian/debian-8.json]: I guess since the salt-jenkins states make sure it's absent, it shouldn't matter and we can just delete it.

Comment (s0undt3ch) [reply AWS/debian/debian-8.json]: Fixed

**Review by bryceml — APPROVED** [review]

**Pull request merged**